pkg/pdatautil/hash_test.go

```diff
@@ -150,6 +150,130 @@ func TestMapHash(t *testing.T) {
+		keys  [][]string
+		equal bool
+	}{
+		{
+			name:  "empty_maps",
+			maps:  []pcommon.Map{pcommon.NewMap(), pcommon.NewMap()},
+			keys:  [][]string{{}, {}},
+			equal: true,
+		},
+		{
+			name: "same_maps_different_order",
+			keys: [][]string{{"k1", "k2"}, {"k1", "k2"}},
+			maps: func() []pcommon.Map {
+				m := []pcommon.Map{pcommon.NewMap(), pcommon.NewMap()}
+				m[0].PutStr("k1", "v1")
+				m[0].PutInt("k2", 1)
+				m[0].PutDouble("k3", 1)
+				m[0].PutBool("k4", true)
+				m[0].PutEmptyBytes("k5").FromRaw([]byte("abc"))
+				sl := m[0].PutEmptySlice("k6")
+				sl.AppendEmpty().SetStr("str")
+				sl.AppendEmpty().SetBool(true)
+				m0 := m[0].PutEmptyMap("k")
+				m0.PutInt("k1", 1)
+				m0.PutDouble("k2", 10)
+
+				m1 := m[1].PutEmptyMap("k")
+				m1.PutDouble("k2", 10)
+				m1.PutInt("k1", 1)
+				m[1].PutEmptyBytes("k5").FromRaw([]byte("abc"))
+				m[1].PutBool("k4", true)
+				sl = m[1].PutEmptySlice("k6")
+				sl.AppendEmpty().SetStr("str")
+				sl.AppendEmpty().SetBool(true)
+				m[1].PutInt("k2", 1)
+				m[1].PutStr("k1", "v1")
+				m[1].PutDouble("k3", 1)
+
+				return m
+			}(),
+			equal: true,
+		},
+		{
+			name: "different_maps_having_same_keys",
+			keys: [][]string{{"k2", "k3"}, {"k2", "k3"}},
+			maps: func() []pcommon.Map {
+				m := []pcommon.Map{pcommon.NewMap(), pcommon.NewMap()}
+				m[0].PutStr("k1", "v1")
+				m[0].PutInt("k2", 1)
+				m[0].PutDouble("k3", 1)
+
+				m[1].PutInt("k2", 1)
+				m[1].PutDouble("k3", 1)
+				m[1].PutDouble("k4", 1)
+
+				return m
+			}(),
+			equal: true,
+		},
+		{
+			name: "same_maps_different_key_order",
+			keys: [][]string{{"k2", "k3"}, {"k3", "k2"}},
+			maps: func() []pcommon.Map {
+				m := []pcommon.Map{pcommon.NewMap(), pcommon.NewMap()}
+				m[0].PutInt("k2", 1)
+				m[0].PutDouble("k3", 1)
+
+				m[1].PutInt("k2", 1)
+				m[1].PutDouble("k3", 1)
+
+				return m
+			}(),
+			equal: false,
+		},
+		{
+			name: "same_maps_with_not_existing_keys",
+			keys: [][]string{{"k10", "k11"}, {"k10", "k11"}},
+			maps: func() []pcommon.Map {
+				m := []pcommon.Map{pcommon.NewMap(), pcommon.NewMap()}
+				m[0].PutStr("k1", "v1")
+
+				m[1].PutInt("k2", 1)
+
+				return m
+			}(),
+			equal: true,
+		},
+		{
+			name: "different_maps",
+			keys: [][]string{{"k2", "k3"}, {"k2", "k3"}},
+			maps: func() []pcommon.Map {
+				m := []pcommon.Map{pcommon.NewMap(), pcommon.NewMap()}
+				m[0].PutInt("k2", 2)
+				m[0].PutDouble("k3", 2)
+
+				m[1].PutInt("k2", 1)
+				m[1].PutDouble("k3", 1)
+
+				return m
+			}(),
+			equal: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			for i := 0; i < len(tt.maps); i++ {
+				for j := i + 1; j < len(tt.maps); j++ {
+					if tt.equal {
+						assert.Equal(t, MapHashSelectedKeys(tt.maps[i], tt.keys[i]), MapHashSelectedKeys(tt.maps[j], tt.keys[j]),
+							"maps %d %v and %d %v must have the same hash, then calculated with keys %v", i, tt.maps[i].AsRaw(), j, tt.maps[j].AsRaw(), tt.keys[i], tt.keys[j])
```

Comment:
Is the second half of the error message useful (`then calculated...`)?  I think from the test name and earlier contents of the error message it would be clear enough. (Same comment for the other assertion here)

If it is valuable, I think it may be missing a `%v`, it looks like there's a mismatch between formatters in the message and arguments to print out.

Reply:
Good catch! Since the order of keys is important now, I've decided to put keys into the error message as well.